multishard_mutation_query.cc

```diff
@@ -841,10 +841,7 @@ class data_query_result_builder {
     stop_iteration consume_end_of_partition()  { return _builder.consume_end_of_partition(); }
     result_type consume_end_of_stream() {
         _builder.consume_end_of_stream();
-        if (_compaction_state.are_limits_reached() || _res_builder->is_short_read()) {
-            return _res_builder->build(_compaction_state.current_full_position());
-        }
-        return _res_builder->build();
+        return _res_builder->build(_compaction_state.current_full_position());
```

Comment:
Doesn't it cost us an extra allocation? For the very common case of only having last pages.

Reply:
Yes, this will copy a partition and clustering key. I think it is worth it to not have to open the rats-nest of some pages having a last pos and others not.

I will check perf/simple_query to assess the impact.

Reply:
> What happens with range tombstones? They can have trailing nulls. Do we encode them correctly?

The type we use for last-pos is `position_in_partition`. It can handle all positions possible with mutation fragments.

> How are pauses after a static row tombstone translated to the client cookie? The client protocol doesn't have the notion of static rows.

Client cookie is just a blob from client's perspective. And since 3131cbea62 we store a `position_in_partition` in it. This reminds me, @nyh can alternator tables have static rows?

Reply:
How does that work? What if a client gets a cookie from a pre-position_in_partition server, and returns it to a post-position_in_partition server?

Reply:
Adding full position-in-partition support was implemented by adding a weight and a region field to `paging_state` in addition to the already present clustering-key. These 3 fields is enough to represent a full position-in-partition.
The cookie is therefore forward compatible but not backwards compatible. An old coordinator would misunderstand a new coordinator's cookie, although in a well defined way. But we have a cluster feature barring the usage of position-in-partition values (range tombstones for example) as last-position in paging cookie.

Reply:
Although we already have replicas providing their last-position, currently they will only stop at rows, stopping at dead rows/range tombstones is controlled by a cluster feature, so we cannot have non at(clustering key) positions in a mixed cluster.

Reply:
Ok, so no problem then.

Reply:
Indeed I see it in paging_state.idl. Good job to all involved.

Reply:
@denesb no, no static rows in Alternator.